src/bin_target.rs

```diff
@@ -38,11 +38,19 @@ pub(crate) fn select_run_binary<'p>(
         let package = metadata
             .packages
             .iter()
-            .find(|package| package.name == *package_name)
+            .find(|package| {
+                // Only consider packages in the current workspace and with the correct name
+                metadata.workspace_members.contains(&package.id) && package.name == *package_name
```

Comment:
The additional filtering is now needed, because we don't pass `--no-deps` to `cargo metadata` anymore (to always only execute it once)